File: python/day16_part2_2018.py

from collections import defaultdict

def addr(registers, A, B, C):
    registers[C] = registers[A] + registers[B]

def addi(registers, A, B, C):
    registers[C] = registers[A] + B

def mulr(registers, A, B, C):
    registers[C] = registers[A] * registers[B]

def muli(registers, A, B, C):
    registers[C] = registers[A] * B

def banr(registers, A, B, C):
    registers[C] = registers[A] & registers[B]

def bani(registers, A, B, C):
    registers[C] = registers[A] & B

def borr(registers, A, B, C):
    registers[C] = registers[A] | registers[B]

def bori(registers, A, B, C):
    registers[C] = registers[A] | B

def setr(registers, A, B, C):
    registers[C] = registers[A]

def seti(registers, A, B, C):
    registers[C] = A

def gtir(registers, A, B, C):
    registers[C] = 1 if A > registers[B] else 0

def gtri(registers, A, B, C):
    registers[C] = 1 if registers[A] > B else 0

def gtrr(registers, A, B, C):
    registers[C] = 1 if registers[A] > registers[B] else 0

def eqir(registers, A, B, C):
    registers[C] = 1 if A == registers[B] else 0

def eqri(registers, A, B, C):
    registers[C] = 1 if registers[A] == B else 0

def eqrr(registers, A, B, C):
    registers[C] = 1 if registers[A] == registers[B] else 0

opcodes = [addr, addi, mulr, muli, banr, bani, borr, bori, setr, seti, gtir, gtri, gtrr, eqir, eqri, eqrr]

def test_opcode(sample, opcode):
    registers = sample[0][:]
    opcode(registers, sample[1][1], sample[1][2], sample[1][3])
    return registers == sample[2]

def parse_input(input_file):
    samples = []
    with open(input_file, 'r') as file:
        while True:
            before_line = file.readline().strip()
            if not before_line:
                break
            before = list(map(int, before_line[9:-1].split(', ')))
            instruction = list(map(int, file.readline().strip().split()))
            after = list(map(int, file.readline().strip()[9:-1].split(', ')))
            samples.append((before, instruction, after))
            file.readline()  # read blank line
    return samples

def part_one(samples):
    count = 0
    for sample in samples:
        num_behave_like = 0
        for opcode in opcodes:
            if test_opcode(sample, opcode):
                num_behave_like += 1
        if num_behave_like >= 3:
            count += 1
    return count

def part_two(samples, program):
    opcode_mapping = {}
    possible_opcodes = defaultdict(set)
    
    for sample in samples:
        opcode_num = sample[1][0]
        possible_opcodes[opcode_num] |= {opcode for opcode in opcodes if test_opcode(sample, opcode)}
    
    while len(opcode_mapping) < 16:
        for opcode_num, opcodes_set in possible_opcodes.items():
            if len(opcodes_set) == 1:
                opcode = opcodes_set.pop()
                opcode_mapping[opcode_num] = opcode
                for other_opcodes in possible_opcodes.values():
                    other_opcodes.discard(opcode)
    
    registers = [0, 0, 0, 0]
    for instruction in program:
        opcode = opcode_mapping[instruction[0]]
        opcode(registers, instruction[1], instruction[2], instruction[3])
    
    return registers[0]

if __name__ == "__main__":
    samples = parse_input("input.txt")
    program = []
    with open("input.txt", 'r') as file:
        for line in file:
            if line[0].isdigit():
                program.append(list(map(int, line.strip().split())))
    
    print(part_one(samples))
    print(part_two(samples, program))

File: nim/day16_part2_2018.nim

import strutils, sequtils, parseutils, tables, sets, os

type
  Registers = array[4, int]
  Instruction = array[4, int]
  Sample = tuple[before: Registers, instruction: Instruction, after: Registers]
  OpcodeProc = proc (r: var Registers, a, b, c: int)

proc addr(r: var Registers, a, b, c: int) = r[c] = r[a] + r[b]
proc addi(r: var Registers, a, b, c: int) = r[c] = r[a] + b
proc mulr(r: var Registers, a, b, c: int) = r[c] = r[a] * r[b]
proc muli(r: var Registers, a, b, c: int) = r[c] = r[a] * b
proc banr(r: var Registers, a, b, c: int) = r[c] = r[a] and r[b]
proc bani(r: var Registers, a, b, c: int) = r[c] = r[a] and b
proc borr(r: var Registers, a, b, c: int) = r[c] = r[a] or r[b]
proc bori(r: var Registers, a, b, c: int) = r[c] = r[a] or b
proc setr(r: var Registers, a, b, c: int) = r[c] = r[a]
proc seti(r: var Registers, a, b, c: int) = r[c] = a
proc gtir(r: var Registers, a, b, c: int) = r[c] = if a > r[b]: 1 else: 0
proc gtri(r: var Registers, a, b, c: int) = r[c] = if r[a] > b: 1 else: 0
proc gtrr(r: var Registers, a, b, c: int) = r[c] = if r[a] > r[b]: 1 else: 0
proc eqir(r: var Registers, a, b, c: int) = r[c] = if a == r[b]: 1 else: 0
proc eqri(r: var Registers, a, b, c: int) = r[c] = if r[a] == b: 1 else: 0
proc eqrr(r: var Registers, a, b, c: int) = r[c] = if r[a] == r[b]: 1 else: 0

const opcodes: array[16, OpcodeProc] = [
  addr, addi, mulr, muli, banr, bani, borr, bori,
  setr, seti, gtir, gtri, gtrr, eqir, eqri, eqrr
]

proc testOpcode(sample: Sample, op: OpcodeProc): bool =
  var currentRegisters = sample.before
  op(currentRegisters, sample.instruction[1], sample.instruction[2], sample.instruction[3])
  return currentRegisters == sample.after

proc parseRegisters(line: string): Registers =
  let parts = line[9 .. ^2].split(", ")
  for i in 0..3:
    result[i] = parseInt(parts[i])

proc parseInstruction(line: string): Instruction =
   let parts = line.split(" ")
   for i in 0..3:
     result[i] = parseInt(parts[i])

proc readInput(filename: string): (seq[Sample], seq[Instruction]) =
  var samples: seq[Sample]
  var program: seq[Instruction]
  var lines = readFile(filename).strip.splitLines
  var i = 0
  while i < lines.len:
    if lines[i].startsWith("Before"):
      let before = parseRegisters(lines[i])
      i += 1
      let instruction = parseInstruction(lines[i])
      i += 1
      let after = parseRegisters(lines[i])
      i += 2 # Skip After line and blank line
      samples.add((before, instruction, after))
    elif lines[i].len > 0 and lines[i][0].isDigit():
       # Assume the rest is the program after the samples and blank lines
       while i < lines.len and lines[i].len > 0:
           program.add(parseInstruction(lines[i]))
           i += 1
       break # Program found and parsed, exit outer loop
    else:
      i += 1 # Skip blank lines between samples or before program
  return (samples, program)

proc partOne(samples: seq[Sample]): int =
  for sample in samples:
    var numBehaveLike = 0
    for op in opcodes:
      if testOpcode(sample, op):
        numBehaveLike += 1
    if numBehaveLike >= 3:
      result += 1

proc partTwo(samples: seq[Sample], program: seq[Instruction]): int =
  var possibleOpcodes = initTable[int, HashSet[OpcodeProc]]()
  var opcodeMapping = initTable[int, OpcodeProc]()

  for sample in samples:
    let opcodeNum = sample.instruction[0]
    var matchingOps = initHashSet[OpcodeProc]()
    for op in opcodes:
      if testOpcode(sample, op):
        matchingOps.incl(op)

    if possibleOpcodes.hasKey(opcodeNum):
        possibleOpcodes[opcodeNum] = possibleOpcodes[opcodeNum] * matchingOps # Intersection
    else:
        possibleOpcodes[opcodeNum] = matchingOps

  # Resolve mappings
  var foundMapping = true
  while opcodeMapping.len < 16 and foundMapping:
      foundMapping = false
      var foundOp: OpcodeProc = nil
      for num, ops in possibleOpcodes.pairs:
          if num notin opcodeMapping and ops.len == 1:
              for op in ops: foundOp = op # Get the single element
              opcodeMapping[num] = foundOp
              foundMapping = true
              # Remove this found op from all other possibilities
              for otherNum in possibleOpcodes.keys:
                  if otherNum != num:
                      possibleOpcodes[otherNum].excl(foundOp)
              break # Restart search for singles after modification


  var registers: Registers # Initialized to zeros
  for instruction in program:
    let op = opcodeMapping[instruction[0]]
    op(registers, instruction[1], instruction[2], instruction[3])

  return registers[0]

when isMainModule:
  let (samples, program) = readInput("input.txt")
  echo partOne(samples)
  echo partTwo(samples, program)

